internal/rpm/files.go

```diff
@@ -0,0 +1,152 @@
+package rpm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime"
+	"strings"
+	"sync"
+	"unique"
+	"weak"
+
+	"github.com/quay/zlog"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/quay/claircore"
+)
+
+// PkgCache is a process-wide cache for [PathSet]s.
+//
+// This allows for all callers to use a single [PathSet] for a given layer.
+var pkgCache fileCache
+
+// FileCache implements a cache of [PathSet] values keyed by the layer digest.
+type fileCache struct {
+	// map[unique.Handle[string]][weak.Pointer[PathSet]]
+	m  sync.Map
+	sf singleflight.Group
+}
+
+func (c *fileCache) cleanupFunc(wp weak.Pointer[PathSet]) func(unique.Handle[string]) {
+	return func(key unique.Handle[string]) {
+		c.m.CompareAndDelete(key, wp)
+	}
+}
+
+// GetPathSet returns a [PathSet] for the provided layer, using a cached
+// version if possible.
+func (c *fileCache) GetPathSet(ctx context.Context, layer *claircore.Layer) (*PathSet, error) {
+	// This makes use of [weak.Pointer], which may be odd at first.
+	//
+	// The weak package has full documentation, but roughly: because a
+	// [weak.Pointer] does not keep a value alive for GC purposes, every use
+	// that retrieves the original pointer needs to assume that it may have
+	// become nil and handle that case before dereferencing it.
+
+	key := unique.Make(layer.Hash.String())
+	if v, ok := c.m.Load(key); ok {
+		if s := v.(weak.Pointer[PathSet]).Value(); s != nil {
+			return s, nil
+		}
+		// Unable to upgrade: it's been garbage collected.
+		// Attempt to delete it so the next caller just has the load fail,
+		// then treat the key as novel.
+		c.m.CompareAndDelete(key, v)
+	}
+
+	ch := c.sf.DoChan(key.Value(), func() (any, error) {
+		set := &PathSet{
```

Comment:
how would you feel about another `Load` over here in case we run into:

thread A calls `Load` on line 48 and doesn't find anything
thread B exits the singleflight and `Store`s the set
thread A now goes into the singleflight and does it all over again since it checked `Load` before B finished

Reply:
Thinking about it some more, I do wonder if we would want to consider the `for` loop pattern instead of the `singleflight` pattern. I do see that's how the `unique` package is implemented

Reply:
The `singleflight` is for work de-duplication, it's not an either-or, I suppose.